runtime/src/main/java/org/corfudb/runtime/collections/streaming/DeltaStream.java

```diff
@@ -155,9 +155,18 @@ public void refresh(StreamAddressSpace streamAddressSpace) {
         }
 
         if (Address.isAddress(streamAddressSpace.getTrimMark())) {
-            Preconditions.checkState(trimMark.get() <= streamAddressSpace.getTrimMark(),
-                    "streamId=%s new trimMark %s has to be >= %s", streamId,
-                    streamAddressSpace.getTrimMark(), trimMark);
+            if (streamAddressSpace.getTrimMark() < trimMark.get()) {
+                // As per Corfu's current design, this check can only be an error message (instead of an actual
+                // validation leading to an exception).
+                // Current 'prefix trim' protocol allows the system to be in a state such that the start address of all
+                // nodes is not the same, i.e., whenever prefixTrim has not gone through all LUs (for instance
+                // consider the case where a TimeoutException hits on the last LU during prefixTrim). If this
+                // LU becomes the new sequencer during this time, its trim mark is expected to regress wrt the other
+                // LUs.
+                // For this error message to become an actual validation it needs to be backed-up by the protocol.
+                log.error("StreamId={} new trimMark {} has to be >= {}", streamId,
```

Comment:
since this condition is expected, should it be logged as debug/info instead of error?

Reply:
mmm to be precise I wouldn't say it's "expected"... in the sense that, there is a code path in which there could be a valid  sequencer regression (the one exposed today, where the new sequencer is bootstrapped from a log unit with failed trims), but there are plenty of other paths where an error could have been introduced, where an "apparent" regression shows up and it's really a bug, for instance, rebuilding bitmaps on log unit restart. So you would need to look into the cause leading to this message (the regression) to see if its legit or not, that's why I'm still leaving it logged in case in the future we want to avoid trim mark regression altogether (a change in the protocol itself). That said, we can probably lower it to warning, as it does not necessarily imply an 'error' but it can't be generalized to 'info' like it's completely fine, as there is no guarantee of the source of the regression.